### DISCUSSION TITLE
gunicorn spends 100% of each request

### DISCUSSION THREAD

**Discussion by freedomwarrior** [netbox-community/netbox#12158] [Help Wanted!]
Hello.
I got fresh install of netbox 3.4.7 + ubuntu 22.04
Some pages opens to long, average ~5s, and in same time gunicorn uses 100% of CPU. 
![image](https://user-images.githubusercontent.com/6192135/229541635-fc722879-dee2-407d-931f-bd4cb6fc17de.png)

**Comment by abhi1693** [comment]
How many objects are you opening per page? The greater the value, the longer it takes to open the page.

**Reply by freedomwarrior** [reply]
Pages are different. It can be login page, or IPAM or something else.

**Reply by candlerb** [reply]
It looks like the specific page you showed a waterfall for is `/ipam/ip-addresses/` which is a *listing* of IP addresses.

What have you set the page length to? (See "Per Page" at the bottom of the table). If the page length is 1000 then 6 seconds doesn't sound too unreasonable - and I'd expect gunicorn to be chomping CPU while it does this.

Try setting the page length to 25 and see if it helps. If not, then further investigation will be required. Please provide a list of the specific URLs where it takes a long time to render.

**Reply by freedomwarrior** [reply]
Login page:
![image](https://user-images.githubusercontent.com/6192135/229723698-111eb24f-c81e-480f-bace-ea689668234a.png)
Providers page(only 2 providers there):
![image](https://user-images.githubusercontent.com/6192135/229723817-56f4258a-88b1-49e7-9ab3-a2be04f4d8de.png)

Pages are different and random.

**Reply by candlerb** [reply]
How sure are you that gunicorn is spinning at 100% for this period? If it wasn't for that, I'd say this is very likely an issue with reverse DNS lookups timing out after 5 seconds.

Is gunicorn actually spinning at 100% of the time, all the time? Is there other load on the system, e.g. API calls?

If the system is mostly idle, and gunicorn is only spinning at 100% while you're fetching the login page, then it might be instructive to attach strace to it. You can then see if it's doing lots of system call activity, or is just calculations.

**Reply by freedomwarrior** [reply]
I monitoring it with top, in idle state I did not see gunicorn in top, at the same time I load random page and this 5 sec I see gunicorn in top with 100% CPU, when the page is loaded - cpu dropped.
But I will check DNS theory and will give feedback about it.

**Reply by candlerb** [reply]
> I monitoring it with top, in idle state I did not see gunicorn in top, at the same time I load random page and this 5 sec I see gunicorn in top with 100% CPU, when the page is loaded - cpu dropped.

Ok, that's good.  So try running this at the shell (as root), *before* hitting the page:

```
strace -ftt -p <pid-of-gunicorn> 2>/var/tmp/strace.out
```

Then hit the page, then when it's finished loading, go back and hit ctrl-C on the strace command. Have a look in the file. Find the timestamps representing the start and end of that 5 second period. What syscalls do you see within that range? In particular, if you see a gap of 5 seconds, then can you show 20 lines either side of that gap?
